### ISSUE TITLE
Opening additional buffers causes error message

### ISSUE THREAD

**Issue by euclio** [rust-lang/rust.vim#48]
Steps to reproduce:

1. Open a rust buffer
2. Open a second rust buffer
    See error message: 
    ```
    Error detected while processing function delimitMate#Get..<SNR>74_get:                                                                  
    line    7:                                                                                                                              
    E716: Key not present in Dictionary: 3.excluded_regions                                                                                 
    E15: Invalid expression: s:options.3.excluded_regions
    ```

3. Open a third buffer
See error message:

    ```
    Error detected while processing function delimitMate#Get..<SNR>74_get:                                                                  
    line    7:                                                                                                                              
    E716: Key not present in Dictionary: 4.excluded_regions                                                                                 
    E15: Invalid expression: s:options.4.excluded_regions
    ```
4. And so on...

**Comment by chris-morgan** [comment]
Sounds like the delimitMate stuff (to make lifetimes work properly) may have ceased to work properly due to upstream changes?

**Comment by euclio** [comment]
Yeah, it looks like an upstream problem stemming from `delimitMate#Get` (is this function considered public? I don't see it documented). Without delving into that code, a simple fix could be to append `,rustLifetimeCandidate,rustGenericLifetimeCandidate` to `delimitMate_excluded_regions` if it exists, else create it with `Comment` (the delimitMate default).

**Comment by euclio** [comment]
@Raimondi any idea what the problem might be?

**Comment by Raimondi** [comment]
I followed the steps but I can't reproduce it. Are you using the latest of delimitMate?

**Comment by euclio** [comment]
Yes. Interesting -- I can reproduce on every computer I use... I'll see if I can put together a minimal vimrc that can reproduce.

**Comment by Raimondi** [comment]
That's a good idea. I'd also try disabling all plugins except for rust.vim and delimitMate to see if another plugin is involved.

**Comment by euclio** [comment]
I am able to reproduce using [vim-plug](http://github.com/junegunn/vim-plug) with the following `.vimrc`:

```viml
call plug#begin('~/.local/share/vim/bundle')

Plug 'Raimondi/delimitMate'

Plug 'rust-lang/rust.vim', { 'for': 'rust' }

call plug#end()
```

Steps to reproduce:

1. `$ vim test.rs`
2. `:sp test2.rs`

**Comment by steveklabnik** [comment]
Those steps do not reproduce for me, using vundle. looks like a vim-plug bug maybe?

**Comment by Raimondi** [comment]
If you want to confirm that vim-plug is guilty of something (not sure what it could be doing) without installing other plugin manager you can modify runtimepath yourself like this:
```viml
set rtp^=~/path/to/rust.vim,~/path/to/delimitMate
```
and comment out the vim-plug related lines.

**Comment by euclio** [comment]
Huh, it looks like it *is* a vim-plug issue. Setting the `rtp` directly works, and removing `{ 'for': 'rust' }` also fixes it.

@junegunn

**Comment by junegunn** [comment]
Interesting problem.

I looked into the issue and found that the error is caused by the assumption that is *usually true* but invalidated by the lazy loading facility of vim-plug. But it's not a fault of vim-plug. You could make rust.vim more robust by making it sure that delimitMate already has touched the buffer before [calling `delimitMate#Get` function](https://github.com/rust-lang/rust.vim/blob/566fa06/ftplugin/rust.vim#L57). Just checking the existence of `g:loaded_delimitMate` is not enough.

Here is the interesting fact; the reason the code currently works is that Vim thinks that `*.rs` files are "hercules" files.

1. While .vimrc is being processed, `filetype plugin indent on` is invoked (note that unlike vundle, vim-plug [does this for you](https://github.com/junegunn/vim-plug/wiki/faq#migrating-from-other-plugin-managers)).
1. `filetype on` part sets up a whole bunch of `BufNewFile` autocmds that assign the correct file type for a new buffer mostly by looking at the file extension. Vim already knows a lot of file extensions (*"rs files are hercules files!"*), but `ftdetect` files from plugins are loaded to register previously unknown extensions or to override obsolete ones (*"no, they are rust files!"*)
1. `filetype plugin` part sets up `FileType` autocmd that loads `ftplugin` files when the type is set.
1. Finished processing .vimrc
1. Plugin files are loaded. delimitMate sets up another `FileType` autocmd to initialize itself whenever a new buffer is open. So this comes **after** the afformentioned `FileType` autocmd.
1. `rs` file is open. Vim first thinks that it's an hercules file, sets the file type accordingly, which triggers `FileType` event.
1. `hercules.vim` is loaded :)
1. delimitMate calls its setup function on this "hercules" file
1. Another `BufNewFile` autocmd from rust.vim is then triggered and it overrides the file type to "rust" triggering another `FileType` event.
1. `ftplugin` from rust.vim is loaded. Luckily, delimitMate already initialized itself during hercules phase, so we don't see the error.

So basically, you got lucky, thanks to hercules. Once Vim decides to remove `au BufNewFile,BufRead *.vc,*.ev,*.rs,*.sum,*.errsum  setf hercules` line from its source code (`filetype.vim`), the code will fail. The reason you see the error when using `for` option of vim-plug is that it allows vim-plug to setup `ftdetect` from rust.vim before the default set of extensions, so no help from hercules.

However, do note that you don't need `for` option in the first place. [As noted in the README page](https://github.com/junegunn/vim-plug#on-demand-loading-of-plugins), it is almost always meaningless for filetype-specific plugins, like rust.vim.

/cc @starcraftman

**Comment by starcraftman** [comment]
@junegunn Interesting bug, thanks for the ping. It feels like something generated by the Heart of Gold's improbability drive given the coincidental chain of events that lead to it working all this time. 

Also, today I learned there exists "hercules" files. I was terribly disappointed that they were just some proprietary circuit checking format. I was expecting something a bit more Olympic. 

As a follow up, seems like the rust devs might want to make a patch request to associate by default `*.rs` with rust. There surely are more rust devs than hercules users.

**Comment by norcalli** [comment]
So what exactly is a solution to this? I have run into this problem myself now.

**Comment by norcalli** [comment]
Nvm, I see that removing `{'for': 'rust'}` fixes this. Indeed an interesting bug.
